tuf/ngclient/updater.py

```diff
@@ -0,0 +1,575 @@
+
+      target_filepath:
+        The path to the target file on the repository. This will be relative to
+        the 'targets' (or equivalent) directory on a given mirror.
+
+    <Exceptions>
+      None.
+
+    <Side Effects>
+      None.
+
+    <Returns>
+      If 'child_role' has been delegated the target with the name
+      'target_filepath', then we return the role name of 'child_role'.
+
+      Otherwise, we return None.
+    """
+
+    child_role_name = child_role.name
+    child_role_paths = child_role.paths
+    child_role_path_hash_prefixes = child_role.path_hash_prefixes
+
+    if child_role_path_hash_prefixes is not None:
+        target_filepath_hash = _get_filepath_hash(target_filepath)
+        for child_role_path_hash_prefix in child_role_path_hash_prefixes:
+            if not target_filepath_hash.startswith(child_role_path_hash_prefix):
+                continue
+
+            return child_role_name
+
+    elif child_role_paths is not None:
+        # Is 'child_role_name' allowed to sign for 'target_filepath'?
+        for child_role_path in child_role_paths:
+            # A child role path may be an explicit path or glob pattern (Unix
+            # shell-style wildcards).  The child role 'child_role_name' is
+            # returned if 'target_filepath' is equal to or matches
+            # 'child_role_path'. Explicit filepaths are also considered
+            # matches. A repo maintainer might delegate a glob pattern with a
+            # leading path separator, while the client requests a matching
+            # target without a leading path separator - make sure to strip any
+            # leading path separators so that a match is made.
+            # Example: "foo.tgz" should match with "/*.tgz".
+            if fnmatch.fnmatch(
+                target_filepath.lstrip(os.sep), child_role_path.lstrip(os.sep)
+            ):
+                logger.debug(
+                    "Child role "
+                    + repr(child_role_name)
+                    + " is allowed to sign for "
+                    + repr(target_filepath)
+                )
+
+                return child_role_name
+
+            logger.debug(
+                "The given target path "
+                + repr(target_filepath)
+                + " does not match the trusted path or glob pattern: "
+                + repr(child_role_path)
+            )
+            continue
+
+    else:
+        # 'role_name' should have been validated when it was downloaded.
+        # The 'paths' or 'path_hash_prefixes' fields should not be missing,
+        # so we raise a format error here in case they are both missing.
+        raise exceptions.FormatError(
+            repr(child_role_name) + " "
+            'has neither a "paths" nor "path_hash_prefixes".  At least'
+            " one of these attributes must be present."
+        )
+
+    return None
+
+
+def _check_file_length(file_object, trusted_file_length):
```

Comment:
Hmm. I get that these are helper functions, but any reason not to fold them into the class?

Reply:
This is issue #1410.

In short: I think the Metadata API should offer these sort of functions (so that there is a "canonical" implementation) and Teodora is currently implementing those in #1437